File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        rMin = 0
        cMin = 0
        rMax = m-1
        cMax = n-1
        ans  = []
        while rMin <= rMax and cMin <= cMax:
            for c in range(cMin,cMax+1):
                ans.append(matrix[rMin][c])
            rMin += 1
            if rMin > rMax:
                break
            for r in range(rMin,rMax+1):
                ans.append(matrix[r][cMax])
            cMax -= 1
            if cMin > cMax:
                break
            for c in range(cMax,cMin-1,-1):
                ans.append(matrix[rMax][c])
            rMax -= 1
            if rMin > rMax:
                break
            for r in range(rMax,rMin-1,-1):
                ans.append(matrix[r][cMin])
            cMin += 1
        return ans

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = (int)matrix.size();
        int n = (int)matrix[0].size();
        vector<int> ans;
        int rMin = 0, rMax = m-1;
        int cMin = 0, cMax = n-1;
        while (rMin <= rMax && cMin <= cMax){
            for (int c=cMin; c<=cMax; c++)
                ans.push_back(matrix[rMin][c]);
            rMin++;
            if (rMin > rMax)
                break;
            for (int r=rMin; r<=rMax; r++)
                ans.push_back(matrix[r][cMax]);
            cMax--;
            if (cMin > cMax)
                break;
            for (int c=cMax; c>=cMin; c--) 
                ans.push_back(matrix[rMax][c]);
            rMax--;
            if (rMin > rMax)
                break;
            for (int r=rMax; r>=rMin; r--)
                ans.push_back(matrix[r][cMin]);
            cMin++;
        }
        return ans;
    }
};